File: common/python/example_connect.py
# Copyright 2016 Zurich Instruments AG

"""
Zurich Instruments LabOne Python API Example

Demonstrate how to connect to a Zurich Instruments device via the Data Server
program.
Create an API session by connecting to a Zurich Instruments
device via the Data Server, ensure the demodulators are enabled and obtain a
single demodulator sample via getSample(). Calculate the sample's RMS
amplitude and add it as a field to the "sample" dictionary.

Note:
  This is intended to be a simple example demonstrating how to connect to a
  Zurich Instruments device from ziPython. In most cases, data acquisition
  should use either ziDAQServer's poll() method or an instance of the
  ziDAQRecorder class, not the getSample() method.

Requirements:
    * LabOne Version >= 20.02
    * Instruments:
        1 x Instrument with demodulators
    * Signal output 1 connected to signal input 1 with a BNC cable.

Usage:
    example_connect.py <device_id>
    example_connect.py -h | --help

Arguments:
    <device_id>  The ID of the device [device_type: .*LI|.*IA|.*IS]

Options:
    -h --help              Show this screen.
    -s --server_host IP    Hostname or IP address of the dataserver [default: localhost]
    -p --server_port PORT  Port number of the data server [default: 8004]

Raises:
    Exception     If the specified devices do not match the requirements.
    RuntimeError  If the devices is not "discoverable" from the API.

See the LabOne Programming Manual for further help:
https://docs.zhinst.com/labone_programming_manual/
"""
import numpy as np
import zhinst.utils


def run_example(
    device_id: str, server_host: str = "localhost", server_port: int = 8004
):
    """run the example."""

    apilevel_example = 6  # The API level supported by this example.
    # Call a zhinst utility function that returns:
    # - an API session `daq` in order to communicate with devices via the data server.
    # - the device ID string that specifies the device branch in the server's node hierarchy.
    (daq, device, _) = zhinst.utils.create_api_session(
        device_id, apilevel_example, server_host=server_host, server_port=server_port
    )
    zhinst.utils.api_server_version_check(daq)

    # Enable the demodulator.
    daq.setInt("/%s/demods/0/enable" % device, 1)
    # Set the demodulator output rate.
    daq.setDouble("/%s/demods/0/rate" % device, 1.0e3)

    # Perform a global synchronisation between the device and the data server:
    # Ensure that the settings have taken effect on the device before issuing
    # the getSample() command.
    daq.sync()

    # Obtain one demodulator sample. If the demodulator is not enabled (as
    # above) then the command will time out: we'll get a RuntimeError showing
    # that a `ZIAPITimeoutException` occured.
    sample = daq.getSample("/%s/demods/0/sample" % device)
    # Calculate the demodulator's magnitude and phase and add them to the sample
    # dict.
    sample["R"] = np.abs(sample["x"] + 1j * sample["y"])
    sample["phi"] = np.angle(sample["x"] + 1j * sample["y"])
    print(f"Measured RMS amplitude is {sample['R'][0]:.3e} V.")


if __name__ == "__main__":
    import sys
    from pathlib import Path

    cli_util_path = Path(__file__).resolve().parent / "../../utils/python"
    sys.path.insert(0, str(cli_util_path))
    cli_utils = __import__("cli_utils")
    cli_utils.run_commandline(run_example, __doc__)
    sys.path.remove(str(cli_util_path))

File: common/matlab/example_connect.m
function r = example_connect(device_id)
% EXAMPLE_CONNECT A simple example demonstrating how to connect to a Zurich Instruments device
%
% USAGE R = EXAMPLE_CONNECT(DEVICE_ID)
%
% Connect to the Zurich Instruments device specified by DEVICE_ID, obtain a
% single demodulator sample and calculate its RMS amplitude R. DEVICE_ID
% should be a string, e.g., 'dev1000' or 'uhf-dev1000'.
%
% NOTE This is intended to be a simple example demonstrating how to connect
% to a Zurich Instruments device from ziPython. In most cases, data acquistion
% should use either ziDAQServer's poll() method or an instance of the
% ziDAQRecorder class.
%
% NOTE Please ensure that the ziDAQ folders 'Driver' and 'Utils' are in your
% Matlab path. To do this (temporarily) for one Matlab session please navigate
% to the ziDAQ base folder containing the 'Driver', 'Examples' and 'Utils'
% subfolders and run the Matlab function ziAddPath().
% >>> ziAddPath;
%
% Use either of the commands:
% >>> help ziDAQ
% >>> doc ziDAQ
% in the Matlab command window to obtain help on all available ziDAQ commands.
%
% See also EXAMPLE_CONNECT_CONFIG, EXAMPLE_POLL.
%
% Copyright 2008-2018 Zurich Instruments AG

clear ziDAQ;

if ~exist('device_id', 'var')
    error(['No value for device_id specified. The first argument to the ' ...
           'example should be the device ID on which to run the example, ' ...
           'e.g. ''dev1000'' or ''uhf-dev1000''.'])
end

% Check the ziDAQ MEX (DLL) and Utility functions can be found in Matlab's path.
if ~(exist('ziDAQ') == 3) && ~(exist('ziCreateAPISession', 'file') == 2)
    fprintf('Failed to either find the ziDAQ mex file or ziDevices() utility.\n')
    fprintf('Please configure your path using the ziDAQ function ziAddPath().\n')
    fprintf('This can be found in the API subfolder of your LabOne installation.\n');
    fprintf('On Windows this is typically:\n');
    fprintf('C:\\Program Files\\Zurich Instruments\\LabOne\\API\\MATLAB2012\\\n');
    return
end

% The API level supported by this example.
supported_apilevel = 6;
% Create an API session; connect to the correct Data Server for the device.
[device, ~] = ziCreateAPISession(device_id, supported_apilevel);
ziApiServerVersionCheck();

branches = ziDAQ('listNodes', ['/' device ], 0);
% Only configure if we have lock-in functionality available.
if any(strcmpi([branches], 'DEMODS'))
  % Enable the demodulator and set a reasonable rate, otherwise the getSample
  % command below will timeout as it won't receive any demodulator data.
  ziDAQ('setInt', ['/' device '/demods/0/enable'], 1);
  ziDAQ('setDouble', ['/' device '/demods/0/rate'], 1.0e3);

  % Perform a global synchronisation between the device and the data server:
  % Ensure that the settings have taken effect on the device before issuing the
  % `getSample` command. Note: `sync` must be issued after waiting for the
  % demodulator filter to settle above.
  ziDAQ('sync');

  % Get a single demodulator sample. Note, `poll` or other higher-level
  % functionality should almost always be be used instead of `getSample`.
  sample = ziDAQ('getSample', ['/' device '/demods/0/sample']);
  r = abs(sample.x + j*sample.y);
  fprintf('Measured RMS amplitude: %eV.\n', r);
else
  r = nan;
end

end
